### PR TITLE
Fix potential lags/deadlocks during tab close

### LINKED ISSUE #14032: Closing tab hangs when using VS Dev tools and Powershell Core
### Windows Terminal version

1.16.2524.0

### Windows build number

 10.0.25201.0

### Other Software

* Visual Studio 2022
* Powershell Core 7.2.6

### Steps to reproduce

1. Create a profile using `pwsh.exe` in Windows Terminal Preview
2. Execute the following lines to add the VS dev tools
```powershell
Import-Module "C:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\Tools\Microsoft.VisualStudio.DevShell.dll"
Enter-VsDevShell 8ee891c5 -SkipAutomaticLocation -DevCmdArguments "-arch=x64 -host_arch=x64"
```
3. Close the tab

### Expected Behavior

Tab closes quickly.

### Actual Behavior

wt locks up entirely doing...something. 

Animation

I've tested this with the release version of Windows Terminal and also regular powershell.exe. Both of those don't exhibit the behavior. It is only with Powershell Core, the most recent release of wt.exe and the Visual Studio dev tools

### PR DESCRIPTION
**Pull request by lhecker**
`ConptyClosePseudoConsole` blocks until OpenConsole exits.
This is problematic for the changes in 666c446, which stopped calling that
function on a background thread to solve a race condition. This commit fixes
the potential lags/deadlocks from waiting on OpenConsole's exit, by adding
`ConptyClosePseudoConsoleNoWait` which only closes the IO handles and allows
OpenConsole to exit naturally. This uncovered another potential deadlock
in `ServiceLocator::RundownAndExit` which might call itself recursively.

Closes #14032

## Validation Steps Performed
* Print tons of text and concurrently close the tab.
  Tab closes, OpenConsole/pwsh exits instantly ✅
* Use `Enter-VsDevShell` and close the tab.
  Tab closes instantly, OpenConsole/pwsh exits after ~5 seconds ✅

### CODE REVIEW

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 1]
```diff
@@ -631,12 +647,6 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
             const auto result{ til::u8u16(std::string_view{ _buffer.data(), read }, _u16Str, _u8State) };
             if (FAILED(result))
             {
-                if (_isStateAtOrBeyond(ConnectionState::Closing))
```
Comment (DHowett) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp:634]: Are we certain that we'll never make it here while a connection is closing? As long as we never transition a connection into the Failed state after we transition it into Closing...

File: src/interactivity/base/ServiceLocator.cpp [hunk 2]
```diff
@@ -39,14 +39,30 @@ void ServiceLocator::SetOneCoreTeardownFunction(void (*pfn)()) noexcept
...
+    }
+    preventRecursion = true;
+
     // MSFT:40146639
     //   The premise of this function is that 1 thread enters and 0 threads leave alive.
     //   We need to prevent anyone from calling us until we actually ExitProcess(),
     //   so that we don't TriggerTeardown() twice. LockConsole() can't be used here,
     //   because doing so would prevent the render thread from progressing.
-    AcquireSRWLockExclusive(&s_shutdownLock);
+    if (locked.exchange(true, std::memory_order_relaxed))
```
Comment (DHowett) [inline src/interactivity/base/ServiceLocator.cpp:62]: Is this safe on all architectures? Is relaxed correct?

File: src/interactivity/base/ServiceLocator.cpp [hunk 3]
```diff
@@ -39,14 +39,30 @@ void ServiceLocator::SetOneCoreTeardownFunction(void (*pfn)()) noexcept
...
+
+    // BODGY:
+    // pRender->TriggerTeardown() might cause another VtEngine pass, which then might fail to write to the IO pipe.
+    // If that happens it calls VtIo::CloseOutput(), which in turn calls ServiceLocator::RundownAndExit().
+    // This prevents the unintended recursion and resulting deadlock.
+    if (preventRecursion)
+    {
+        return;
+    }
+    preventRecursion = true;
```
Comment (DHowett) [inline src/interactivity/base/ServiceLocator.cpp:55]: yikes. could we use an `std::exchange` in the if condition?

Comment (DHowett) [reply src/interactivity/base/ServiceLocator.cpp:55]: mostly "yikes" because of the manual implementation of a recursive lock bail-out

File: src/interactivity/base/ServiceLocator.cpp [hunk 4]
```diff
@@ -39,14 +39,30 @@ void ServiceLocator::SetOneCoreTeardownFunction(void (*pfn)()) noexcept
...
+
     // MSFT:40146639
     //   The premise of this function is that 1 thread enters and 0 threads leave alive.
     //   We need to prevent anyone from calling us until we actually ExitProcess(),
     //   so that we don't TriggerTeardown() twice. LockConsole() can't be used here,
     //   because doing so would prevent the render thread from progressing.
-    AcquireSRWLockExclusive(&s_shutdownLock);
+    if (locked.exchange(true, std::memory_order_relaxed))
+    {
+        Sleep(INFINITE);
```
Comment (DHowett) [inline src/interactivity/base/ServiceLocator.cpp:64]: So just to confirm.

If we get here, it is the first entry and we take the lock, we proceed .We might cause either...

1. A recursive call to `RundownAndExit`
2. Another thread to `RundownAndExit`

The first one is fixed by `preventRecursion`, and the second one is fixed by `locked` -> `Sleep(INFINITE)`?

What if we trigger another thread to `RundownAndExit` while we're waiting on it to complete? Is that possible?

File: src/interactivity/base/ServiceLocator.cpp [hunk 5]
```diff
@@ -39,14 +39,30 @@ void ServiceLocator::SetOneCoreTeardownFunction(void (*pfn)()) noexcept
     s_oneCoreTeardownFunction = pfn;
 }
 
-[[noreturn]] void ServiceLocator::RundownAndExit(const HRESULT hr)
+void ServiceLocator::RundownAndExit(const HRESULT hr)
```
Comment (DHowett) [inline src/interactivity/base/ServiceLocator.cpp:42]: Changes here are scary. Are we sure this won't cause the OneCore console to regress?

**Comment by DHowett** [comment]
What behavior do you see for process management when you open a hundred tabs and then close the window (f.ex)?

File: src/cascadia/TerminalConnection/ConptyConnection.cpp (hunk 1 above)
Comment (lhecker) [reply src/cascadia/TerminalConnection/ConptyConnection.cpp:634]: That's actually entirely possible and the source of the race condition that I fixed.
While `_isStateAtOrBeyond` itself might be atomic, that doesn't mean it protects us from other threads changing the state under us concurrently. Unless I misunderstood your question, I don't believe that this is any more or less correct than before.

File: src/interactivity/base/ServiceLocator.cpp (hunk 2 above)
Comment (lhecker) [reply src/interactivity/base/ServiceLocator.cpp:62]: Yes it's safe on all architectures. Atomic operations themselves are always exactly as atomic as any other, no matter what memory order is specified. Relaxed only means that no additional memory is synchronized/flushed (which is fine here).

File: src/interactivity/base/ServiceLocator.cpp (hunk 5 above)
Comment (lhecker) [reply src/interactivity/base/ServiceLocator.cpp:42]: No I'm not sure, but I don't believe so. The only effective difference is the addition of the `thread_local` and I could revert the change to `s_shutdownLock` if you'd like.

**Comment by lhecker** [comment]
> What behavior do you see for process management when you open a hundred tabs and then close the window (f.ex)?

I opened ~40 pwsh tabs and it took about 2-3s for WT to disappear. The pwsh/OpenConsole processes however quit instantly. (Maybe some sort of WinUI lag? Didn't investigate it any further.)
Closing ~20 tabs of the VS dev shell made WT disappear after 1-2s. The pwsh/OpenConsole processes got detached and disappeared after 5-10s (some were faster, some took slightly longer).

**Review by DHowett — APPROVED** [review]
I'm okay with this. I'll pack it into a 1.16 build and ship it out for folks...!

File: src/winconpty/ft_pty/ConPtyTests.cpp [hunk 6]
```diff
@@ -382,7 +382,7 @@ void ConPtyTests::DiesOnClose()
     Log::Comment(NoThrowString().Format(L"Sleep a bit to let the process attach"));
     Sleep(100);
 
-    _ClosePseudoConsoleMembers(&pty);
+    _ClosePseudoConsoleMembers(&pty, TRUE);
```
Comment (carlos-zamora) [inline src/winconpty/ft_pty/ConPtyTests.cpp:385]: Y'all don't like default params, huh? :(

File: src/interactivity/inc/ServiceLocator.hpp [hunk 7]
```diff
@@ -86,7 +86,6 @@ namespace Microsoft::Console::Interactivity
 
         static HWND LocatePseudoWindow(const HWND owner = nullptr /*HWND_DESKTOP = 0*/);
 
-    protected:
```
Comment (carlos-zamora) [inline src/interactivity/inc/ServiceLocator.hpp:89]: Why remove protected? Should we keep protected for some of these?

File: src/interactivity/base/ServiceLocator.cpp [hunk 8]
```diff
@@ -39,14 +39,29 @@ void ServiceLocator::SetOneCoreTeardownFunction(void (*pfn)()) noexcept
...
+
     // MSFT:40146639
     //   The premise of this function is that 1 thread enters and 0 threads leave alive.
     //   We need to prevent anyone from calling us until we actually ExitProcess(),
     //   so that we don't TriggerTeardown() twice. LockConsole() can't be used here,
     //   because doing so would prevent the render thread from progressing.
-    AcquireSRWLockExclusive(&s_shutdownLock);
+    if (locked.exchange(true, std::memory_order_relaxed))
+    {
+        Sleep(INFINITE);
```
Comment (carlos-zamora) [inline src/interactivity/base/ServiceLocator.cpp:63]: Please put a big ol' comment here why it's ok to put `Sleep(INFINITE);`. I think it'd be nice instead of having to find this PR and read through the discussion.

**Review by carlos-zamora — COMMENTED** [review]
ngl, it's a bit hard to follow. I think with a few answered questions I'll feel comfortable approving though.

File: src/winconpty/ft_pty/ConPtyTests.cpp (hunk 6 above)
Comment (lhecker) [reply src/winconpty/ft_pty/ConPtyTests.cpp:385]: It's a C function - that's why it can't have that.

File: src/interactivity/inc/ServiceLocator.hpp (hunk 7 above)
Comment (lhecker) [reply src/interactivity/inc/ServiceLocator.hpp:89]: The class is `final` so that `protected` has no effect. It's also not usual / best practice to make classes any more private than necessary. A non-public copy constructor for instance is pretty unusual since you would only be able to use them inside the class. It gets even more unusual when you consider that the functions are explicitly deleted, so not even the class itself can use them. In short, this code was kinda whacky. 😅

File: src/winconpty/ft_pty/ConPtyTests.cpp (hunk 6 above)
Comment (DHowett) [reply src/winconpty/ft_pty/ConPtyTests.cpp:385]: Also FWIW no, I personally don't like default parameters! I also don't like underdocumented boolean parameters for reasons I've mentioned before, but since this is an internal/in-file interface I don't care as much.

File: src/interactivity/base/ServiceLocator.cpp (hunk 8 above)
Comment (lhecker) [reply src/interactivity/base/ServiceLocator.cpp:65]: It ended up being a bit of a shorter comment. But I think that one, plus the longer explanation in front of the `if` condition should sufficiently explain what this does... 🤔

**Review by carlos-zamora — APPROVED** [review]
